src/Type/Doctrine/DBAL/QueryBuilder/QueryBuilderExecuteMethodExtension.php

```diff
@@ -0,0 +1,51 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Doctrine\DBAL\QueryBuilder;
+
+use Doctrine\DBAL\Driver\ResultStatement;
+use Doctrine\DBAL\Query\QueryBuilder;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Identifier;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\DynamicMethodReturnTypeExtension;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+
+class QueryBuilderExecuteMethodExtension implements DynamicMethodReturnTypeExtension
+{
+
+	public function getClass(): string
+	{
+		return QueryBuilder::class;
+	}
+
+	public function isMethodSupported(MethodReflection $methodReflection): bool
+	{
+		return $methodReflection->getName() === 'execute';
+	}
+
+	public function getTypeFromMethodCall(MethodReflection $methodReflection, MethodCall $methodCall, Scope $scope): Type
+	{
+		$defaultReturnType = ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+
+		$var = $methodCall->var;
+		while ($var instanceof MethodCall) {
+			$nameObject = $var->name;
```

Comment:
One more thing here - we're not checking that `$var->var` is still `Doctrine\DBAL\Query\QueryBuilder`. You might have this situation. The best way to do that is early return in case:

```php
$queryBuilderType = new ObjectType(QueryBuilder::class); // this line should be above the while loop
$varType = $scope->getType($var->var);
if (!$queryBuilderType->isSuperTypeOf($varType)->yes()) {
    return $defaultReturnType;
}
```

Once this is resolved, I'll merge and release it :)

Reply:
done :heavy_check_mark: